### PR TITLE
fix: correctly handle unions in `verify_type`

### LINKED ISSUE #577: Traceback when "Union[Member, User]" is used as type hint in slash command parameters
### Summary

When "Union[Member, User]" is used as a type hint in a slash command parameter, a traceback from disnake is shown when the slash command is invoked targeting a user who should resolve to a User object, not a Member object..

### Reproduction Steps

Reference the below Minimal Reproducible Code and invoke the relevant slash command against a user who is not currently in the server. This issue does *not* occur when the target of the slash command is in the server (and resolves to a Member object).

### Minimal Reproducible Code

```python
@discord_bot.slash_command()
async def whois(
    inter: ApplicationCommandInteraction,
    user: Union[Member, User] = commands.Param(
        description="The member to display information about"
    ),
) -> None:
    pass
```

### Expected Results

The slash command should be invoked successfully without a traceback when targeting a user who is not in the server. This used to work in v2.4.0 of disnake - not positive if it is broken in v2.5.0 as well.

### Actual Results

The following traceback is observed when this slash command is invoked:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/disnake/ext/commands/interaction_bot_base.py", line 1264, in process_application_commands
    await app_command.invoke(interaction)
  File "/usr/local/lib/python3.8/site-packages/disnake/ext/commands/slash_core.py", line 680, in invoke
    await call_param_func(self.callback, inter, self.cog, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/disnake/ext/commands/params.py", line 811, in call_param_func
    kwargs[param.param_name] = await param.convert_argument(
  File "/usr/local/lib/python3.8/site-packages/disnake/ext/commands/params.py", line 464, in convert_argument
    return await self.verify_type(inter, argument)
disnake.ext.commands.errors.MemberNotFound: Member "redacted_valid_snowflake_here" not found.
```

### Intents

default, members, message_content

### System Information

```markdown
$ python -m disnake -v
- Python v3.8.10-final
- disnake v2.5.1-final
    - disnake pkg_resources: v2.5.1
- aiohttp v3.7.4.post0
- system info: Linux 5.4.0-120-generic #136-Ubuntu SMP Fri Jun 10 13:40:48 UTC 2022
```
```


### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

Talked with Mari about this over Discord, she said she'd open up an issue for this. I opened this up to save her some time 😅

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Fixes #577.

An annotation like `Union[User, Member]` passes the `issubclass_(self.type, disnake.Member)` check since `issubclass_` explicitly handles unions; the following `isinstance(argument, disnake.Member)` check did not accept `User` objects, even though they should've been permitted based on the annotation type.
This PR improves the validation logic to correctly handle cases like this as well.

Regression from #283 - third test case of `test_verify_type` checks for this particular one.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by onerandomusername — CHANGES_REQUESTED** [review]

File: disnake/ext/commands/params.py [hunk 1]
```diff
@@ -443,14 +440,22 @@ async def get_default(self, inter: ApplicationCommandInteraction) -> Any:
         return default
 
     async def verify_type(self, inter: ApplicationCommandInteraction, argument: Any) -> Any:
-        """Check if a type of an argument is correct and possibly fix it"""
-        if issubclass_(self.type, disnake.Member):
-            if isinstance(argument, disnake.Member):
-                return argument
+        """Check if the type of an argument is correct and possibly raise if it's not"""
```
Comment (onerandomusername) [inline disnake/ext/commands/params.py:443]: style: needs a `.`

Comment (shiftinv) [reply disnake/ext/commands/params.py]: c00fbe7a0d956c54fa3eb54110bb0c62b9e2f2ca

File: disnake/ext/commands/params.py [hunk 2]
```diff
@@ -443,14 +440,22 @@ async def get_default(self, inter: ApplicationCommandInteraction) -> Any:
...
+        if self.discord_type.value not in (6, 9):
```
[the remark this answers is not in the document]
Comment (shiftinv) [reply disnake/ext/commands/params.py]: Benchmarked a few alternatives, the one in 5cedadf80cfa84c1b4088fc297776cd7400cf535 seems to be the fastest one, even being slightly faster than the old check.

**Review by onerandomusername — CHANGES_REQUESTED** [review]
This still has a minor bug with users and members:
image
```py
    @parent.sub_command("user")
    async def user(self, itr: disnake.CommandInteraction, user: disnake.User):
        await itr.send(type(user))
```

**Comment by shiftinv** [comment]
> This still has a minor bug with users and members

~~it's not a bug, it's a feature~~
But yeah, that seems to be working as intended, even if `Member` doesn't inherit from `User` it ends up having the same attributes+methods due to `@flatten_user`

**Comment by onerandomusername** [comment]
Sounds good, then.

**Review by onerandomusername — APPROVED** [review]

**Pull request merged**